test/integration_test.py

```diff
@@ -404,6 +410,78 @@ def _get_entities_by_project(self, entity_type, project_short_name):
             params['search_after_uid'] = pagination['search_after_uid']
         return entities
 
+    def test_concurrent_portal_db_crud(self):
+        """
+        Use multithreading to simulate multiple users simultaneously modifying
+        the portals database.
+        """
+        if config.deployment_stage not in ('sandbox', 'nadove'):
+            log.info('Skipping portal CRUD test to avoid polluting database')
+            return
+
+        # Currently takes about 50 seconds and creates a 25 kb db file.
+        n_threads = 10
+        n_ops = 10
+        portal_service = PortalService()
+
+        # Assign to local variable outside of callback so we don't need to put
+        # the entire test inside of _crud call
+        portal_db = None
+
+        def get_db(db):
+            # `make format` is convinced this needs to be on two lines.
+            nonlocal \
+                portal_db
+            portal_db = list(db)
+
+        def add_entry(db: JSONs, thread_name: str, count: int):
+            mock_entry = {
+                "portal_id": "foo",
+                "integrations": [
+                    {
+                        "integration_id": "bar",
+                        "entity_type": "project",
+                        "integration_type": "get",
+                        "entity_ids": ["baz"]
+                    }
+                ],
+                "mock-count": f"{thread_name}-{count}"
+            }
+
+            return list(db) + [cast(JSON, mock_entry)]
+
+        class TestThread(Thread):
+
+            def __init__(self, *args, **kwargs):
+                super().__init__(*args, **kwargs)
+                self.portal_service = PortalService()
+
+            def run(self):
+                for i in range(n_ops):
+                    self.portal_service._crud(lambda db: add_entry(db, self.name, i))
+
+        portal_service._crud(get_db)
+        init_db = cast(JSONs, portal_db)
+
+        threads = [TestThread(name=f'T{i}') for i in range(n_threads)]
+        for thread in threads:
+            thread.start()
+
+        while any(thread.is_alive() for thread in threads):
```

Comment:
Traditionally, one would use `Thread.join()` which is blocking and avoids the busy waiting.